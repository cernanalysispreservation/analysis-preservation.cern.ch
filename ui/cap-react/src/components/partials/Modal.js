import React, { useRef, useEffect } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import styled from "styled-components";
import { AiOutlineClose } from "react-icons/ai";

const Wrapper = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 200;
  background: rgba(0, 0, 0, 0.5);
`;

const Content = styled.div`
  border-radius: 3px;
  min-width: 320px;
  max-width: 90%;
  height: ${props => props.position.height};
  overflow: scroll;
  background: ${props => props.background};
  display: flex;
  flex-direction: column;
  overflow: auto;

  position: absolute;
  top: ${props => props.position.top};
  left: ${props => props.position.left};
  right: ${props => props.position.right};
  bottom: ${props => props.position.bottom};
  transform: ${props => props.position.transform};
`;

const Header = styled.div`
  display: grid;
  grid-template-columns: 50px 1fr 50px;
  padding: 10px 15px;
  border-bottom: ${props =>
    props.separator ? "0.1px solid rgba(0, 0, 0, 0.3)" : ""};
  color: #000;
`;

const Body = styled.div`
  padding: ${props => props.padding};
`;

const Heading = styled.div`
  grid-column: 2/3;
  justify-self: center;
  align-self: center;
`;

const CloseButton = styled.div`
  grid-column: 3/4;
  justify-self: end;
  align-self: center;
  display: flex;
  align-items: center;
  padding: 5px;
  &:hover {
    background: rgba(228, 230, 235, 1);
    border-radius: 50%;
    cursor: pointer;
  }
`;
const Modal = ({
  Tag = "h3",
  onClose = null,
  children,
  padding = "10px",
  title = "",
  separator = false,
  position = "center",
  background = "rgb(255,255,255)",
  full = false
}) => {
  const modal = useRef(null);

  useEffect(() => {
    document.addEventListener("keydown", escFunction, false);
    return () => {
      document.removeEventListener("keydown", escFunction, false);
    };
  }, []);

  const escFunction = event => {
    // when the user clicks the ESC button
    // the menu will hide
    if (event.keyCode === 27) {
      onClose();
    }
  };

  // useEffect in order to capture clicks that are not in the menu
  useEffect(
    () => {
      function handleClickOutside(event) {
        const dom = ReactDOM.findDOMNode(modal.current);
        if (modal.current && !dom.contains(event.target)) {
          onClose();
        }
      }
      // Bind the event listener
      document.addEventListener("mousedown", handleClickOutside);
      return () => {
        // Unbind the event listener on clean up
        document.removeEventListener("mousedown", handleClickOutside);
      };
    },
    [modal]
  );

  const getPositionByName = name => {
    const choices = {
      left: {
        top: 0,
        left: 0,
        height: full ? "100%" : null
      },
      right: {
        top: 0,
        right: 0,
        height: full ? "100%" : null
      },
      center: {
        top: "50%",
        left: "50%",
        transform: "translate(-50%, -50%)"
      }
    };
    return choices[name];
  };
  return (
    <Wrapper id="cap-modal">
      <Content
        ref={modal}
        position={getPositionByName(position)}
        background={background}
      >
        <Header separator={separator}>
          <Heading>
            <Tag>{title}</Tag>
          </Heading>
          <CloseButton onClick={onClose} className="func-close">
            <AiOutlineClose size={25} color="rgba(0,0,0,0.6)" />
          </CloseButton>
        </Header>
        <Body padding={padding}>{children}</Body>
      </Content>
    </Wrapper>
  );
};

Modal.propTypes = {
  Tag: PropTypes.oneOf(["h1", "h2", "h3", "h4", "h5", "h6"]),
  onClose: PropTypes.func,
  children: PropTypes.node,
  padding: PropTypes.string,
  title: PropTypes.string,
  separator: PropTypes.bool,
  background: PropTypes.string,
  position: PropTypes.string,
  full: PropTypes.bool
};

export default Modal;
