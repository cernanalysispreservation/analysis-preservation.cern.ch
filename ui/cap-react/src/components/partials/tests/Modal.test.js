import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { configure, shallow, mount } from "enzyme";

import Modal from "../Modal";

configure({ adapter: new Adapter() });

// set the wrapper for the modal
let wrapper;
let content = <h1>MyTitle</h1>;
beforeEach(() => {
  wrapper = mount(
    <Modal title="This is my title" Tag="h5" onClose={() => console.log()}>
      {content}
    </Modal>
  );
});

// clear and unmount after the end of each test
afterEach(() => {
  wrapper.unmount();
});

describe("Modal Suite", () => {
  it("Snapshot image", () => {
    const shallowWrapper = shallow(<Modal />);
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("Modal Children render", () => {
    // children content
    expect(wrapper.find("h1")).toHaveLength(1);
    // header tag
    expect(wrapper.find("h5")).toHaveLength(1);
    // overrides defaults header tag value
    expect(wrapper.find("h3")).toHaveLength(0);
  });

  it("Modal close icon click functionality", () => {
    const spy = jest.spyOn(console, "log");
    const btn = wrapper.find("div.func-close");
    expect(spy).toHaveBeenCalledTimes(0);
    btn.simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
