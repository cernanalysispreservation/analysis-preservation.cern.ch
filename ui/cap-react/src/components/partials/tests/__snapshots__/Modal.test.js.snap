// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Suite Snapshot image 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modal />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Unknown
        background="rgb(255,255,255)"
        full={false}
        position={
                Object {
                        "left": "50%",
                        "top": "50%",
                        "transform": "translate(-50%, -50%)",
                      }
        }
>
        <Unknown
                separator={false}
        >
                <Unknown>
                        <h3 />
                </Unknown>
                <Unknown
                        className="func-close"
                        onClick={null}
                >
                        <AiOutlineClose
                                color="rgba(0,0,0,0.6)"
                                size={25}
                        />
                </Unknown>
        </Unknown>
        <Unknown
                padding="10px"
        />
</Unknown>,
      "id": "cap-modal",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "background": "rgb(255,255,255)",
        "children": Array [
          <Unknown
            separator={false}
>
            <Unknown>
                        <h3 />
            </Unknown>
            <Unknown
                        className="func-close"
                        onClick={null}
            >
                        <AiOutlineClose
                                    color="rgba(0,0,0,0.6)"
                                    size={25}
                        />
            </Unknown>
</Unknown>,
          <Unknown
            padding="10px"
/>,
        ],
        "full": false,
        "position": Object {
          "left": "50%",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Unknown>
                <h3 />
</Unknown>,
              <Unknown
                className="func-close"
                onClick={null}
>
                <AiOutlineClose
                                color="rgba(0,0,0,0.6)"
                                size={25}
                />
</Unknown>,
            ],
            "separator": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <h3 />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                },
                "ref": null,
                "rendered": "",
                "type": "h3",
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-ifAKCX",
                  "isStatic": false,
                  "lastClassName": "eDacmc",
                  "rules": Array [
                    "
  grid-column: 2/3;
  justify-self: center;
  align-self: center;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-ifAKCX",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <AiOutlineClose
                  color="rgba(0,0,0,0.6)"
                  size={25}
/>,
                "className": "func-close",
                "onClick": null,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "color": "rgba(0,0,0,0.6)",
                  "size": 25,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": false,
                  "lastClassName": "hwGnud",
                  "rules": Array [
                    "
  grid-column: 3/4;
  justify-self: end;
  align-self: center;
  display: flex;
  align-items: center;
  padding: 5px;
  &:hover {
    background: rgba(228, 230, 235, 1);
    border-radius: 50%;
    cursor: pointer;
  }
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": false,
              "lastClassName": "bXFFEE",
              "rules": Array [
                "
  display: grid;
  grid-template-columns: 50px 1fr 50px;
  padding: 10px 15px;
  border-bottom: ",
                [Function],
                ";
  color: #000;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": undefined,
            "padding": "10px",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": false,
              "lastClassName": "hIfEnT",
              "rules": Array [
                "
  padding: ",
                [Function],
                ";
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": false,
          "lastClassName": "kaDPzj",
          "rules": Array [
            "
  border-radius: 3px;
  min-width: 320px;
  max-width: 90%;
  max-height: ",
            [Function],
            ";
  height: ",
            [Function],
            ";
  overflow: scroll;
  background: ",
            [Function],
            ";
  display: flex;
  flex-direction: column;
  overflow: auto;

  position: absolute;
  top: ",
            [Function],
            ";
  left: ",
            [Function],
            ";
  right: ",
            [Function],
            ";
  bottom: ",
            [Function],
            ";
  transform: ",
            [Function],
            ";
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": false,
        "lastClassName": "laammi",
        "rules": Array [
          "
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 200;
  background: rgba(0, 0, 0, 0.5);
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Unknown
          background="rgb(255,255,255)"
          full={false}
          position={
                    Object {
                              "left": "50%",
                              "top": "50%",
                              "transform": "translate(-50%, -50%)",
                            }
          }
>
          <Unknown
                    separator={false}
          >
                    <Unknown>
                              <h3 />
                    </Unknown>
                    <Unknown
                              className="func-close"
                              onClick={null}
                    >
                              <AiOutlineClose
                                        color="rgba(0,0,0,0.6)"
                                        size={25}
                              />
                    </Unknown>
          </Unknown>
          <Unknown
                    padding="10px"
          />
</Unknown>,
        "id": "cap-modal",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "background": "rgb(255,255,255)",
          "children": Array [
            <Unknown
              separator={false}
>
              <Unknown>
                            <h3 />
              </Unknown>
              <Unknown
                            className="func-close"
                            onClick={null}
              >
                            <AiOutlineClose
                                          color="rgba(0,0,0,0.6)"
                                          size={25}
                            />
              </Unknown>
</Unknown>,
            <Unknown
              padding="10px"
/>,
          ],
          "full": false,
          "position": Object {
            "left": "50%",
            "top": "50%",
            "transform": "translate(-50%, -50%)",
          },
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Unknown>
                  <h3 />
</Unknown>,
                <Unknown
                  className="func-close"
                  onClick={null}
>
                  <AiOutlineClose
                                    color="rgba(0,0,0,0.6)"
                                    size={25}
                  />
</Unknown>,
              ],
              "separator": false,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <h3 />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "",
                  },
                  "ref": null,
                  "rendered": "",
                  "type": "h3",
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-ifAKCX",
                    "isStatic": false,
                    "lastClassName": "eDacmc",
                    "rules": Array [
                      "
  grid-column: 2/3;
  justify-self: center;
  align-self: center;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-ifAKCX",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <AiOutlineClose
                    color="rgba(0,0,0,0.6)"
                    size={25}
/>,
                  "className": "func-close",
                  "onClick": null,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "color": "rgba(0,0,0,0.6)",
                    "size": 25,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": false,
                    "lastClassName": "hwGnud",
                    "rules": Array [
                      "
  grid-column: 3/4;
  justify-self: end;
  align-self: center;
  display: flex;
  align-items: center;
  padding: 5px;
  &:hover {
    background: rgba(228, 230, 235, 1);
    border-radius: 50%;
    cursor: pointer;
  }
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": false,
                "lastClassName": "bXFFEE",
                "rules": Array [
                  "
  display: grid;
  grid-template-columns: 50px 1fr 50px;
  padding: 10px 15px;
  border-bottom: ",
                  [Function],
                  ";
  color: #000;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": undefined,
              "padding": "10px",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bxivhb",
                "isStatic": false,
                "lastClassName": "hIfEnT",
                "rules": Array [
                  "
  padding: ",
                  [Function],
                  ";
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bxivhb",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": false,
            "lastClassName": "kaDPzj",
            "rules": Array [
              "
  border-radius: 3px;
  min-width: 320px;
  max-width: 90%;
  max-height: ",
              [Function],
              ";
  height: ",
              [Function],
              ";
  overflow: scroll;
  background: ",
              [Function],
              ";
  display: flex;
  flex-direction: column;
  overflow: auto;

  position: absolute;
  top: ",
              [Function],
              ";
  left: ",
              [Function],
              ";
  right: ",
              [Function],
              ";
  bottom: ",
              [Function],
              ";
  transform: ",
              [Function],
              ";
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "lastClassName": "laammi",
          "rules": Array [
            "
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 200;
  background: rgba(0, 0, 0, 0.5);
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
